File: piper/android/app/src/main/java/io/gopiper/piper/engine/statements/lists_sort.kt
package io.gopiper.piper.engine.statements

import android.util.Log
import io.gopiper.piper.engine.Piper
import org.json.JSONObject
import java.lang.Exception

suspend fun Piper.executeListsSort(statement: JSONObject, onError: (e: Exception) -> Unit): List<Any?>? {
    try {
        Log.d("hello", "execute_lists_sort")
        val pointer = statement.getJSONObject("pointer")
        val list = executeStatement(pointer["list"].toString()) as? ArrayList<Any?> ?: return null
        val isReversed = pointer["direction"].toString() == "-1"

        return when (pointer["type"].toString()) {
            "NUMERIC" -> {
                val s = list.sortedBy { it.toString().toDouble() }
                if (isReversed) return s.reversed()
                return s
            }
            "TEXT" -> {
                val s = list.sortedBy { it.toString() }
                if (isReversed) return s.reversed()
                return s
            }
            "IGNORE_CASE" -> {
                val s = list.sortedBy { it.toString().lowercase() }
                if (isReversed) return s.reversed()
                return s
            }
            else -> null
        }
    } catch (e: Exception) {
        onError(e)
        return null
    }
}

File: piper/extensions/chrome/src/engine/statements/lists_sort.js
import { stringify } from "../util"



function listsGetSortCompare(type, direction) {
    var compareFuncs = {
      'NUMERIC': function(a, b) {
          return Number(a) - Number(b); },
      'TEXT': function(a, b) {
          return a.toString() > b.toString() ? 1 : -1; },
      'IGNORE_CASE': function(a, b) {
          return a.toString().toLowerCase() > b.toString().toLowerCase() ? 1 : -1; },
    };
    var compare = compareFuncs[type];
    return function(a, b) { return compare(a, b) * direction; };
}


export default async function(statement, onError) {
    try {
        let pointer = statement.pointer
        let list = await this.executeStatement(stringify(pointer.list))
        let direction = pointer.direction
        let sortType = stringify(pointer.type)
       
        return list.slice().sort(listsGetSortCompare(sortType, direction))
    } catch (e) {
        onError(e)
    }
}
